### PR TITLE
ParseJSON middleware

### PR DESCRIPTION
**Pull request by lmeunier**
This middleware makes it much easier to write JSON based API, by
popuplating the environment map (c.Env) with the decoded JSON from the
body request.

### CODE REVIEW

**Comment by mkrull** [comment]
@lmeunier What is the actual advantage of unmarshaling into `interface{}`/`map[string]interface{}` instead of delaying decoding until the target struct type is known? Especially with nested JSON having the `interface{}` and unpacking manually seems like a disadvantage to me. Is there an easy way to cast `map[string]interface{}` into a specific struct type?

**Comment by lmeunier** [comment]
In the current version, the ``ParseJSON`` middleware is not aware of the struct to use to unmarshal the incoming data, so it can only unmarshal into an ``interface{}``. I agree with you, this is not very useful actually...

It would be nice if we can pass the target struct to the middleware so it can unmarshal into a struct rather than an ``interface{}``. But I don't know how to do it, any idea?

**Comment by ekanna** [comment]
@lmeunier Have a look at this gojson https://github.com/ChimeraCoder/gojson

**Comment by parkr** [comment]
Generally :+1: on this one.

**Comment by mkrull** [comment]
@ekanna gojson does not work on runtime and to get a type definition from json is not the goal of this middleware also it would not be useful to generate a type definition/struct at runtime in that case anyway because using a type unknown at compile time would be basically as useful as having `interface{}` which we already get with the current solution.

I have been thinking about this some time because I would also like to see this middleware happen but it is hard to imagine a clean generic way of getting automatic deserialization working in a middleware.

**Comment by basvanbeek** [comment]
If people see a need for an automatic JSON deserializer in a middleware that's great and all but I really wonder how much gain it provides over using a simple decoder function like this:

```
func DecodeJSON(r *http.Request, o interface{}) (error) {
	if r.ContentLength < 1 {
		return errors.New("Invalid or empty payload")
	}
	raw := make([]byte, r.ContentLength)
	_, err := io.ReadFull(r.Body, raw)
	if err == nil {
		err = json.Unmarshal(raw, &o)
	}
	return err
}
```

Now depending on *what* you expect to receive on a particular end node (usually I need specific parameters in a specific format) you can either provide something like this:

```
func Login (c web.C, w http.ResponseWriter, r *http.Request) {
	var p struct {
		User string
		Pass string
	}
	if err := utils.DecodeJSON(r, &p); err != nil {
		/* invalid payload */
	}
	... test parameter(s) sanity and do stuff...
}
```

or if you want to go completely freeform just provide an interface{} and inspect types...

Bottom line is when building an API you usually have specific input parameters in a specific format which often are known at compile time so why not use that in your node handler functions instead of having a generic middleware handle it and push it into an interface{} that is much more tedious to inspect in your node handler functions.

**Comment by zenazn** [comment]
To circle back on this, most of the Go web apps I've seen have done something very similar to what @basvanbeek proposed. I think there's a lot of room for improving the abstractions here, but I'm not convinced that the middleware proposed in this PR is that.

Unless anyone strongly objects, I'm going to close out this PR to keep things tidy.
